src/audio/drc/drc_generic.c

```diff
@@ -0,0 +1,435 @@
+
+#include <sof/audio/component.h>
+#include <sof/audio/drc/drc.h>
+#include <sof/audio/drc/drc_math.h>
+#include <sof/audio/format.h>
+#include <stdint.h>
+
+/* This is the knee part of the compression curve. Returns the output level
+ * given the input level x. */
+static float knee_curveK(struct sof_drc_params *p, float x)
+{
+	const float knee_alpha = Q_CONVERT_QTOF(p->knee_alpha, 20);
+	const float knee_beta = Q_CONVERT_QTOF(p->knee_beta, 30);
+	const float K = Q_CONVERT_QTOF(p->K, 12);
+
+	/* The formula in knee_curveK is linear_threshold +
+	 * (1 - expf(-k * (x - linear_threshold))) / k
+	 * which simplifies to (alpha + beta * expf(gamma))
+	 * where alpha = linear_threshold + 1 / k
+	 *	 beta = -expf(k * linear_threshold) / k
+	 *	 gamma = -k * x
+	 */
+	return knee_alpha + knee_beta * knee_expf(-K * x);
+}
+
+/* Full compression curve with constant ratio after knee. Returns the ratio of
+ * output and input signal. */
+static float volume_gain(struct sof_drc_params *p, float x)
+{
+	const float knee_threshold = Q_CONVERT_QTOF(p->knee_threshold, 20);
+	const float linear_threshold = Q_CONVERT_QTOF(p->linear_threshold, 20);
+	const float ratio_base = Q_CONVERT_QTOF(p->ratio_base, 30);
+	const float slope = Q_CONVERT_QTOF(p->slope, 20);
+	float y;
+
+	if (x < knee_threshold) {
+		if (x < linear_threshold)
+			return 1;
+		y = knee_curveK(p, x) / x;
+	} else {
+		/* Constant ratio after knee.
+		 * log(y/y0) = s * log(x/x0)
+		 * => y = y0 * (x/x0)^s
+		 * => y = [y0 * (1/x0)^s] * x^s
+		 * => y = ratio_base * x^s
+		 * => y/x = ratio_base * x^(s - 1)
+		 * => y/x = ratio_base * e^(log(x) * (s - 1))
+		 */
+		y = ratio_base * knee_expf(logf(x) * (slope - 1));
+	}
+
+	return y;
+}
+
+/* Update detector_average from the last input division. */
+static void drc_update_detector_average(struct drc_state *state, struct sof_drc_params *p, int nch)
+{
+	float abs_input_array[DRC_DIVISION_FRAMES];
+	const float sat_release_frames_inv_neg = Q_CONVERT_QTOF(p->sat_release_frames_inv_neg, 20);
+	const float sat_release_rate_at_neg_two_db = Q_CONVERT_QTOF(p->sat_release_rate_at_neg_two_db, 20);
+	float detector_average = Q_CONVERT_QTOF(state->detector_average, 30);
+	int div_start, i, ch;
+	float sample;
+
+	/* Calculate the start index of the last input division */
+	if (state->pre_delay_write_index == 0) {
+		div_start = DRC_MAX_PRE_DELAY_FRAMES - DRC_DIVISION_FRAMES;
+	} else {
+		div_start = state->pre_delay_write_index - DRC_DIVISION_FRAMES;
+	}
+
+	/* The max abs value across all channels for this frame */
+	for (i = 0; i < DRC_DIVISION_FRAMES; i++) {
+		abs_input_array[i] = 0.0f;
+		for (ch = 0; ch < nch; ch++) {
+			sample = Q_CONVERT_QTOF(state->pre_delay_buffers[ch][div_start + i], 31);
+			abs_input_array[i] = fmaxf(abs_input_array[i], fabsf(sample));
+		}
+	}
+
+	for (i = 0; i < DRC_DIVISION_FRAMES; i++) {
+		/* Compute compression amount from un-delayed signal */
+		float abs_input = abs_input_array[i];
```

Comment:
Variables introductions have been in function start. It's probably to see immediately how much the function uses stack.